sycl/doc/SYCL_compiler_and_runtime_design.md

```diff
@@ -0,0 +1,259 @@
+does the following:
+- creates a .sycl_offloading.descriptor symbol which is a structure containing
+the number of device images and the array of the device images themselves
+```C++
+struct __tgt_device_image {
+  void *ImageStart;
+  void *ImageEnd;
+};
+struct __tgt_bin_desc {
+  int32_t NumDeviceImages;
+  __tgt_device_image *DeviceImages;
+};
+__tgt_bin_desc .sycl_offloading.descriptor;
+```
+- creates a `void .sycl_offloading.descriptor_reg()` function and registers it
+for execution at module loading; this function invokes the `__tgt_register_lib`
+registration function (the name can also be specified via an option) which must
+be implemented by the runtime and which registers the device images with the
+runtime:
+```C++
+void __tgt_register_lib(__tgt_bin_desc *desc);
+```
+
+- creates a `void .sycl_offloading.descriptor_unreg()` function and registers
+it for execution at module unloading; this function calls the
+`__tgt_unregister_lib` function (the name can also be specified via an option)
+which must be implemented by the runtime and which unregisters the device
+images with the runtime:
+```C++
+void __tgt_unregister_lib(__tgt_bin_desc *desc);
+```
+
+Once the offload wrapper object file is ready, the driver finally invokes the
+host linker giving it the following input:
+- all the application host objects (result of compilation or unbundling)
+- the offload wrapper object file
+- all the host libraries needed by the application
+- the SYCL runtime library
+The result is so-called "fat binary image" containing the host code, code for
+all the targets plus the registration/unregistration functions and the
+information about the device binary images.
+
+When compilation and linking is done in single compiler driver invocation, the
+bundling and unbundling steps are skipped.
+
+*Design note: the described scheme differs from current llvm.org
+implementation.  Current design uses Linux-specific linker script approach and
+requires that all the linked fat objects are compiled for the same set of
+targets. The described design uses OS-neutral offload-wrapper tool and does not
+impose restrictions on fat objects.*
+
+### Integration with SPIR-V format
+*TBD*
+
+### Address spaces handling
+SYCL 1.2.1 language defines several address spaces where data can reside, the
+same as OpenCL - global, local, private and constant. From the spec: "In OpenCL
+C, these address spaces are manually specified using OpenCL-specific keywords.
+In SYCL, the device compiler is expected to auto-deduce the address space for
+pointers in common situations of pointer usage. However, there are situations
+where auto-deduction is not possible".
+
+We believe that requirement for the compiler to automatically deduce address
+spaces is too strong. Instead the following approach will be implemented in the
+compiler:
+
+*TBD*
+
+### Compiler/Runtime interface
+
+## SYCL Runtime architecture
+*TBD*
+
+Supported extensions Intel subgroups
```

Comment:
Probably 

```
## Supported extensions

* Intel subgroups [cl_intel_subgroups](https://www.khronos.org/registry/OpenCL/extensions/intel/cl_intel_subgroups.html)
```

Reply:
Probably it should a link to the SYCL extension.
I'll add it once it available.